src/Controller/Controller.php

```diff
@@ -504,15 +507,22 @@ public function invokeAction()
                 'plugin' => $request->getParam('plugin'),
             ]);
         }
+
         /** @var callable $callable */
         $callable = [$this, $request->getParam('action')];
 
         $result = $callable(...array_values($request->getParam('pass')));
-        if ($result instanceof Response) {
-            $this->response = $result;
+        if ($result === null) {
+            return $result;
+        }
+
+        if (!$result instanceof ResponseInterface) {
+            throw new UnexpectedValueException(
+                'Controller actions can only return ResponseInterface instance or null.'
```

Comment:
Should we tell the user the type of thing we got that was unexpected?

Reply:
Done